ambry-cloud/src/main/java/com.github.ambry.cloud/CloudFindToken.java

```diff
@@ -0,0 +1,130 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud;
+
+import com.github.ambry.store.FindToken;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Objects;
+
+
+/**
+ * FindToken implementation used by the {@link CloudBlobStore}.
+ */
+public class CloudFindToken implements FindToken {
+
+  static final short VERSION_0 = 0;
+  static final short CURRENT_VERSION = VERSION_0;
+  private final short version;
+  private final long latestUploadTime;
+  private final String latestBlobId;
+  private final long bytesRead;
+
+  /** Constructor for start token */
+  public CloudFindToken() {
+    this(0, null, 0);
+  }
+
+  /** Constructor for in-progress token */
+  public CloudFindToken(long latestUploadTime, String latestBlobId, long bytesRead) {
+    this.version = CURRENT_VERSION;
+    this.latestUploadTime = latestUploadTime;
+    this.latestBlobId = latestBlobId;
+    this.bytesRead = bytesRead;
+  }
+
+  /**
+   * Utility to construct a new CloudFindToken from a previous instance and the results of a findEntriesSince query.
+   * @param prevToken the previous CloudFindToken.
+   * @param queryResults the results of a findEntriesSince query.
+   * @return the updated token.
+   */
+  public static CloudFindToken getUpdatedToken(CloudFindToken prevToken, List<CloudBlobMetadata> queryResults) {
+    if (queryResults.isEmpty()) {
+      return prevToken;
+    } else {
+      CloudBlobMetadata lastResult = queryResults.get(queryResults.size() - 1);
+      long bytesReadThisQuery = queryResults.stream().mapToLong(CloudBlobMetadata::getSize).sum();
+      return new CloudFindToken(lastResult.getUploadTime(), lastResult.getId(), prevToken.getBytesRead() + bytesReadThisQuery);
+    }
+  }
+
+  @Override
+  public byte[] toBytes() {
+    byte[] buf = null;
+    switch (version) {
+      case VERSION_0:
+        int size = Short.BYTES + 2 * Long.BYTES;
+        buf = new byte[size];
+        ByteBuffer bufWrap = ByteBuffer.wrap(buf);
+        // add version
+        bufWrap.putShort(version);
+        // add latestUploadTime
+        bufWrap.putLong(latestUploadTime);
+        // add bytesRead
+        bufWrap.putLong(bytesRead);
+        if (latestBlobId != null) {
+          bufWrap.putShort((short) latestBlobId.length());
+          bufWrap.put(latestBlobId.getBytes());
+        } else {
+          bufWrap.putShort((short) 0);
+        }
+        break;
+      default:
+        throw new IllegalStateException("Unknown version: " + version);
+    }
+    return buf;
+  }
+
+  @Override
+  public long getBytesRead() {
+    return bytesRead;
+  }
+
+  public String getLatestBlobId() {
+    return latestBlobId;
+  }
+
+  public long getLatestUploadTime() {
+    return latestUploadTime;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
+1 This is a good example of a good implementation of `hashCode` and `equals` since it makes good use of `Objects.equals` (which is good for checking equality with objects that may be null) and `Objects.hash` (which outsources the hard parts of developing a good hash with multiple fields) and I like the use of `getClass()` instead of `instanceof` since it demands an exact class match instead of subtypes being ok

Reply:
I used the Intellij code generator so can't claim too much credit :)